airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/partition_generator.py

```diff
@@ -0,0 +1,20 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from abc import ABC, abstractmethod
+from typing import Iterable
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+
+
+class PartitionGenerator(ABC):
+    @abstractmethod
+    def generate(self, sync_mode: SyncMode) -> Iterable[Partition]:
+        """
+        Generates partitions for a given sync mode.
+        :param sync_mode: SyncMode
+        :return:
+        """
+        pass
```

Comment:
What about cases when don't know the number of partitions ahead? For example when partitioning depends on the result of the previous request?